src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,349 @@
+package scala                     //TODO: Change this strawman --> we just need sizeHintIfCheap from scala
+package collection.mutable
+
+import scala.collection.{GenSeq, generic, mutable}
+import scala.reflect.ClassTag
+
+/** An implementation of a double-ended queue that internally uses a resizable circular buffer
+  *  Append, prepend, removeFirst, removeLast and random-access (indexed-lookup and indexed-replacement)
+  *  take amortized constant time. In general, removals and insertions at i-th index are O(min(i, n-i))
+  *  and thus insertions and removals from end/beginning are fast.
+  *
+  *  @author  Pathikrit Bhowmick
+  *  @version 2.12
+  *  @since   2.12
+  *
+  *  @tparam A  the type of this ArrayDeque's elements.
+  *
+  *  @define Coll `mutable.ArrayDeque`
+  *  @define coll array deque
+  *  @define thatinfo the class of the returned collection. In the standard library configuration,
+  *    `That` is always `ArrayDeque[B]` because an implicit of type `CanBuildFrom[ArrayDeque, B, ArrayDeque[B]]`
+  *    is defined in object `ArrayDeque`.
+  *  @define bfinfo an implicit value of class `CanBuildFrom` which determines the
+  *    result class `That` from the current representation type `Repr`
+  *    and the new element type `B`. This is usually the `canBuildFrom` value
+  *    defined in object `ArrayDeque`.
+  *  @define orderDependent
+  *  @define orderDependentFold
+  *  @define mayNotTerminateInf
+  *  @define willNotTerminateInf
+  */
+@SerialVersionUID(1L)
+class ArrayDeque[A] private[ArrayDeque](
+    private[ArrayDeque] var array: Array[AnyRef],
+    private[ArrayDeque] var start: Int,
+    private[ArrayDeque] var end: Int
+) extends mutable.AbstractBuffer[A]
+    with mutable.Buffer[A]
+    with generic.GenericTraversableTemplate[A, ArrayDeque]
+    with mutable.BufferLike[A, ArrayDeque[A]]
+    with mutable.IndexedSeq[A]
+    with mutable.IndexedSeqOptimized[A, ArrayDeque[A]]
+    with mutable.Builder[A, ArrayDeque[A]]
+    with Serializable {
```

Comment:
It would be nice to have a comment somewhere in here describing the invariants that `ArrayDeque` is supposed to maintain.  Off the top of my head, `array` has length of a power of two, `mask` is equal to `array.length-1` (actually, do we need to store this in a field?  A single decrement is super-fast), and `start` and `end` are...actually...I'm not sure?